### PR TITLE
docs: add field decorators migration documentation

### PR DESCRIPTION
**Pull request by kapunahelewong**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
Add new section to deprecation doc about field decorators migration.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: aio/content/guide/deprecations.md [hunk 1]
```diff
@@ -561,3 +561,103 @@ The following table shows all methods that the migration maps from `Renderer` to
...
+}
+
+@Directive({...})
+class Dir extends Base {
+}
+```
+
+After:
+```typescript
+@Directive({...}) // adds @Directive()
```
[the remark this answers is not in the document]
Comment (devversion) [reply aio/content/guide/deprecations.md]: Yes. You're right that it will be just an empty `@Directive()` decorator.

File: aio/content/guide/deprecations.md [hunk 2]
```diff
@@ -561,3 +561,103 @@ The following table shows all methods that the migration maps from `Renderer` to
...
 |`animate(element, startingStyles, keyframes, duration, delay, easing, previousPlayers?)`|Throws an error (same behavior as `Renderer.animate()`)|
+
+
+## Undecorated directives (field decorators)
+
+<!-- Removing ngBaseDef by migrating classes -->
+
+### What does this schematic do?
+
+This schematic scans classes in the compilation and if a class is not decorated at the top level, but has Angular decorated fields, adds `@Directive()` to the class.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/deprecations.md]: Added! Thank you for the copy. It's great!

File: aio/content/guide/deprecations.md [hunk 3]
```diff
@@ -561,3 +561,103 @@ The following table shows all methods that the migration maps from `Renderer` to
...
+
+Angular decorated fields include:
+- `@Input()`
+- `@Output()`
+- `@HostBinding()`
+- `@HostListener()`
+- `@ViewChild()` / `@ViewChildren()`
+- `@ContentChild()` / `@ContentChildren()`
+
+It does not include param decorators for dependency injection, such as `@Host()`, `@SkipSelf()`, `@Optional()`, and `@Self()`.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/deprecations.md]: Fixed.

File: aio/content/guide/deprecations.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/deprecations.md]: Got them all. Thank you both. :)

File: aio/content/guide/deprecations.md [hunk 4]
```diff
@@ -561,3 +561,103 @@ The following table shows all methods that the migration maps from `Renderer` to
...
+```
+
+
+However, in Ivy this model won't compile as there is no annotation on `Base` which would generate an Ivy definition to carry the information about inputs, outputs, queries, or host bindings. Ivy always requires a `@Directive()` or `@Component()` selector for any class using Angular features.
+
+Ivy removes the need for `ngBaseDef` by requiring that every undecorated class with decorated fields has an `@Directive()` annotation, and thus receives an `ngDirectiveDef` which carries the information about those fields. This change simplifies the code, reduces the runtime's complexity, and allows removal of the code supporting `ngBaseDef` logic.
+
+### When should I add a @Directive() decorator without a selector?
+
+If you're using dependency injection, or any Angular-specific
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/deprecations.md]: Ah, thank you. I was reaching there.

File: aio/content/guide/migration-field-decorators.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+}
+```
+
+However, in Ivy this model won't compile as there is no
+decorator on `Base` which would generate an Ivy definition
+to carry the information about inputs, outputs, queries, or
+host bindings. Ivy always requires a `@Directive()`
+or `@Component()` selector for any class using Angular features.
+
+Ivy removes the need for `ngBaseDef` by requiring that
```
Comment (kara) [inline aio/content/guide/migration-field-decorators.md]: Base def is actually an Ivy-only concept. We will be removing it, but its removal will occur from one Ivy version to a later Ivy version. Regardless, I think defs are too low-level for this guide, so let's remove this section.

I think the only thing we need to say - once we've merged this into the other guide - is that there were two motivations for extending the requirement for a @Directive decorator to include field decorators: 

1 - the former distinction was arbitrary. Why does DI require a decorator but other Angular features don't? It's hard to remember which is which. It's easier if we can just say "If you use any Angular feature (whether it's a field decorator or DI), you must have a class decorator"

2 - this change allows us to reduce the size of the framework code. Making Angular features work in undecorated base classes was complicated and expensive, code-wise. If we ask users to add the required decorators, the framework can be smaller for everyone.

**Review by kara — CHANGES_REQUESTED** [review]
TLDR: let's take the info here and merge it into the existing guide.

**Comment by kapunahelewong** [comment]
Have added the info here to https://github.com/angular/angular/pull/33069 because it fit better in  the undecorated classes migration doc. Closing this PR. Please refer to #33069 for the latest on this topic.
